Inline small wasm functions

justinmichaud opened the pull request:
#### fa0d2ca34a9cdfc451719663ed074136e63ac3a0
<pre>
Inline small wasm functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=252717">https://bugs.webkit.org/show_bug.cgi?id=252717</a>

Reviewed by Yusuke Suzuki.

Support inlining small wasm functions. When we decide to inline,
we parse the inlinee and add its b3 nodes directly to our graph.

Our inlinee&apos;s exception handlers are added directly, but they take care to add extra
buffer state for their catch handlers. This state is ignored in the caller.

Arguments and return values are handled transparently, and we skip the
prologue, epilogue or stack check as the caller should handle these things.

Finally, we add a new testing mode that aggressively inlines to test this new pass.

* JSTests/wasm/stress/simple-inline-exception.js: Added.
(catch):
* JSTests/wasm/stress/simple-inline-exception.wasm: Added.
* JSTests/wasm/stress/simple-inline-exception.wat: Added.
* JSTests/wasm/stress/simple-inline.js: Added.
* JSTests/wasm/stress/simple-inline.wasm: Added.
* JSTests/wasm/stress/simple-inline.wat: Added.
* Source/JavaScriptCore/b3/B3Generate.cpp:
(JSC::B3::generateToAir):
* Source/JavaScriptCore/b3/B3StackmapValue.cpp:
(JSC::B3::StackmapValue::dumpMeta const):
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::generateCompilerConstructionSite):
(JSC::B3::Value::deepDump const):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::restoreCalleeSavesFromEntryFrameCalleeSavesBuffer):
(JSC::AssemblyHelpers::restoreCalleeSavesFromVMEntryFrameCalleeSavesBufferImpl):
(JSC::AssemblyHelpers::emitSave):
(JSC::AssemblyHelpers::emitRestore):
(JSC::AssemblyHelpers::emitSaveCalleeSavesFor):
(JSC::AssemblyHelpers::emitRestoreCalleeSavesFor):
* Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
(JSC::AssemblyHelpers::Spooler::execute):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::throwException):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addStackMap):
(JSC::Wasm::B3IRGenerator::takeStackmaps):
(JSC::Wasm::B3IRGenerator::takeExceptionHandlers):
(JSC::Wasm::B3IRGenerator::B3IRGenerator):
(JSC::Wasm::B3IRGenerator::addLoop):
(JSC::Wasm::B3IRGenerator::preparePatchpointForExceptions):
(JSC::Wasm::B3IRGenerator::emitCatchImpl):
(JSC::Wasm::B3IRGenerator::addDelegateToUnreachable):
(JSC::Wasm::B3IRGenerator::addReturn):
(JSC::Wasm::B3IRGenerator::emitInlineDirectCall):
(JSC::Wasm::B3IRGenerator::addCall):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::Callee::Callee):
(JSC::Wasm::Callee::runWithDowncast):
* Source/JavaScriptCore/wasm/WasmCompilationMode.cpp:
(JSC::Wasm::makeString):
* Source/JavaScriptCore/wasm/WasmCompilationMode.h:
(JSC::Wasm::isOSREntry):
(JSC::Wasm::isAnyBBQ):
(JSC::Wasm::isAnyOMG):
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::emitThrowImpl):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::functionWasmSizeImportSpace const):
(JSC::Wasm::ModuleInformation::functionWasmSize const):
(JSC::Wasm::ModuleInformation::isSIMDFunctionImportSpace const):
(JSC::Wasm::ModuleInformation::isSIMDFunction const):
(JSC::Wasm::ModuleInformation::addSIMDFunction):
(JSC::Wasm::ModuleInformation::doneSeeingFunction):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/261191@main">https://commits.webkit.org/261191@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2520991e5dcbe25a500f64206dd79f628ed00dd8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/110867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/19951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/2218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/114828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/21374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/11085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/2218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/116619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/21374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/21374 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/103231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/99494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/10645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13100 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/11085 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/100599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/18501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/100599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/15045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/108631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4234 "Built successfully and passed tests") | | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/108631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -3969,6 +4113,61 @@ B3::PatchpointValue* B3IRGenerator::createTailCallPatchpoint(BasicBlock* block,
+    unsigned inlineDepth = 0;
+    Checked<uint32_t> inlinedBytes = 0;
+    for (auto* currentFrame = this; currentFrame; currentFrame = currentFrame->m_inlineParent) {
+        ++inlineDepth;
```
Let's record depth in each frame, so `m_inlineDepth + 1` is enough here.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -3969,6 +4113,61 @@ B3::PatchpointValue* B3IRGenerator::createTailCallPatchpoint(BasicBlock* block,
+    Checked<uint32_t> inlinedBytes = 0;
+    for (auto* currentFrame = this; currentFrame; currentFrame = currentFrame->m_inlineParent) {
+        ++inlineDepth;
+        inlinedBytes += currentFrame->m_inlinedBytes;
```
Isn't it only computing inlinedbytes of this callstack frame? We should compute this globally.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -81,6 +83,9 @@ inline void Callee::runWithDowncast(const Func& func)
         break;
+    case CompilationMode::OMGForInlineCallMode:
+        RELEASE_ASSERT_NOT_REACHED();
+        break;
```
Let's not add this. `CompilationMode` is for Callee. If the compiler is not creating special callee for that, then we should not add it, this is not the data type for that.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -3969,6 +4113,61 @@ B3::PatchpointValue* B3IRGenerator::createTailCallPatchpoint(BasicBlock* block,
+    BasicBlock* continuation = m_proc.addBlock();
+
+    const FunctionData& function = m_info.functions[calleeFunctionIndex];
+    m_protectedInlineeGenerators.constructAndAppend(*this, CompilationMode::OMGForInlineCallMode, calleeFunctionIndex, continuation, WTFMove(getArgs));
```
Instead of creating `CompilationMode::OMGForInlineCallMode`, we should just pass the current CompilationMode, and separately adding a boolean flag etc. to indicate that this is inlined calleframe.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -805,11 +833,21 @@ class B3IRGenerator {
+    Vector<Value*> m_inlinedArgs;
+    Vector<Variable*> m_inlinedResults;
+    unsigned m_inlineDepth { 0 };
+    Checked<uint32_t> m_inlinedBytes = { 0 };
```
Not use `=`. Align it to WebKit's style.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -3969,6 +4109,57 @@ B3::PatchpointValue* B3IRGenerator::createTailCallPatchpoint(BasicBlock* block,
+{
+    ASSERT(!m_inlinedBytes || !m_inlineRoot);
+    return m_inlineDepth < Options::maximumWasmDepthForInlining()
+        && (m_inlineRoot ? m_inlineRoot : this)->m_inlinedBytes.value() < Options::maximumWasmCallerSizeForInlining();
```
Why not always set `m_inlineRoot`?

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -4099,6 +4290,17 @@ auto B3IRGenerator::addCall(uint32_t functionIndex, const TypeDefinition& signat
+        && canInline()) {
+        dataLogLnIf(WasmB3IRGeneratorInternal::verboseInlining, " inlining call to ", functionIndex - m_numImportFunctions, " from ", m_functionIndex);
+        RELEASE_ASSERT(!m_info.callCanClobberInstance(functionIndex));
+        (m_inlineRoot ? m_inlineRoot : this)->m_inlinedBytes += m_info.functionWasmSizeImportSpace(functionIndex);
```
Ditto. Let's always set `m_inlineRoot` and simplify this update.

Constellation reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -3671,7 +3786,7 @@ auto B3IRGenerator::addDelegateToUnreachable(ControlType& target, ControlType& d
         targetDepth = target.tryDepth();
 
-    m_exceptionHandlers.append({ HandlerType::Delegate, data.tryStart(), ++m_callSiteIndex, 0, m_tryCatchDepth, targetDepth });
+    m_exceptionHandlers.append({ HandlerType::Delegate, data.tryStart(), advanceCallSiteIndex(), 0, m_tryCatchDepth, targetDepth });
```
Are we collecting exception handlers of inlined callee?

Constellation replied:
Also can you review other fields to ensure that we are collecting them? Next week, we should make IRBuilder instance one, and we should separate data structures to CompileCallFrame to ensure this and do this cleanly.

justinmichaud replied:
Yes

Commit:
Inline small wasm functions
https://bugs.webkit.org/show_bug.cgi?id=252717

Reviewed by Yusuke Suzuki.

Support inlining small wasm functions. When we decide to inline,
we parse the inlinee and add its b3 nodes directly to our graph.

Our inlinee's exception handlers are added directly, but they take care to add extra
buffer state for their catch handlers. This state is ignored in the caller.

Arguments and return values are handled transparently, and we skip the
prologue, epilogue or stack check as the caller should handle these things.

Finally, we add a new testing mode that aggressively inlines to test this new pass.

* JSTests/wasm/stress/simple-inline-exception.js: Added.
(catch):
* JSTests/wasm/stress/simple-inline-exception.wasm: Added.
* JSTests/wasm/stress/simple-inline-exception.wat: Added.
* JSTests/wasm/stress/simple-inline.js: Added.
* JSTests/wasm/stress/simple-inline.wasm: Added.
* JSTests/wasm/stress/simple-inline.wat: Added.
* Source/JavaScriptCore/b3/B3Generate.cpp:
(JSC::B3::generateToAir):
* Source/JavaScriptCore/b3/B3StackmapValue.cpp:
(JSC::B3::StackmapValue::dumpMeta const):
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::generateCompilerConstructionSite):
(JSC::B3::Value::deepDump const):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::restoreCalleeSavesFromEntryFrameCalleeSavesBuffer):
(JSC::AssemblyHelpers::restoreCalleeSavesFromVMEntryFrameCalleeSavesBufferImpl):
(JSC::AssemblyHelpers::emitSave):
(JSC::AssemblyHelpers::emitRestore):
(JSC::AssemblyHelpers::emitSaveCalleeSavesFor):
(JSC::AssemblyHelpers::emitRestoreCalleeSavesFor):
* Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
(JSC::AssemblyHelpers::Spooler::execute):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::throwException):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addStackMap):
(JSC::Wasm::B3IRGenerator::takeStackmaps):
(JSC::Wasm::B3IRGenerator::takeExceptionHandlers):
(JSC::Wasm::B3IRGenerator::B3IRGenerator):
(JSC::Wasm::B3IRGenerator::addLoop):
(JSC::Wasm::B3IRGenerator::preparePatchpointForExceptions):
(JSC::Wasm::B3IRGenerator::emitCatchImpl):
(JSC::Wasm::B3IRGenerator::addDelegateToUnreachable):
(JSC::Wasm::B3IRGenerator::addReturn):
(JSC::Wasm::B3IRGenerator::emitInlineDirectCall):
(JSC::Wasm::B3IRGenerator::addCall):
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::Callee::Callee):
(JSC::Wasm::Callee::runWithDowncast):
* Source/JavaScriptCore/wasm/WasmCompilationMode.cpp:
(JSC::Wasm::makeString):
* Source/JavaScriptCore/wasm/WasmCompilationMode.h:
(JSC::Wasm::isOSREntry):
(JSC::Wasm::isAnyBBQ):
(JSC::Wasm::isAnyOMG):
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::emitThrowImpl):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::functionWasmSizeImportSpace const):
(JSC::Wasm::ModuleInformation::functionWasmSize const):
(JSC::Wasm::ModuleInformation::isSIMDFunctionImportSpace const):
(JSC::Wasm::ModuleInformation::isSIMDFunction const):
(JSC::Wasm::ModuleInformation::addSIMDFunction):
(JSC::Wasm::ModuleInformation::doneSeeingFunction):
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/261191@main
